--- ddclient	2015-02-23 00:51:18.000000000 +0100
+++ ddclient_final_2015-03-17	2015-03-18 01:18:01.000000000 +0100
@@ -471,6 +471,15 @@
      'warned-min-interval'       => setv(T_ANY,    0, 1, 0, 0,             undef),
      'warned-min-error-interval' => setv(T_ANY,    0, 1, 0, 0,             undef),
     },
+    'dinahosting-service-common-defaults'       => {
+     'server'              => setv(T_FQDNP,  1, 0, 1, 'dinahosting.com', undef),
+     'script'              => setv(T_STRING, 0, 1, 1, '/special/api.php', undef),
+     'min-interval'        => setv(T_DELAY,  0, 0, 1, interval('30s'),     0),
+     'max-interval'        => setv(T_DELAY,  0, 0, 1, interval('25d'),     0),
+     'min-error-interval'  => setv(T_DELAY,  0, 0, 1, interval('8m'),      0),
+     'warned-min-interval'       => setv(T_ANY,    0, 1, 0, 0,             undef),
+     'warned-min-error-interval' => setv(T_ANY,    0, 1, 0, 0,             undef),
+    },
     'dnsexit-service-common-defaults'       => {
      'server'              => setv(T_FQDNP,  1, 0, 1, 'www.dnsexit.com', undef),
      'script'              => setv(T_STRING, 0, 1, 1, '/RemoteUpdate.sv', undef),
@@ -722,6 +731,15 @@
                  $variables{'service-common-defaults'},
                   ),
     },
+    'dinahosting' => {
+     'updateable' => undef,
+     'update'     => \&nic_dinahosting_update,
+     'examples'   => \&nic_dinahosting_examples,
+     'variables'  => merge(
+                 $variables{'dinahosting-service-common-defaults'},
+                 $variables{'service-common-defaults'},
+                  ),
+    },
     'direktdomains' => {
      'updateable' => undef,
      'update'     => \&nic_direktdomains_update,
@@ -4491,6 +4509,109 @@
 EoEXAMPLE
 }
 ######################################################################
+## nic_dinahosting_update
+######################################################################
+sub nic_dinahosting_update {
+     debug("\nnic_dinahosting_update -------------------");
+
+     ## update each configured host
+     my $replyCheck = undef;
+     foreach my $h (@_) {
+          my $ip = delete $config{$h}{'wantip'};
+          info("setting IP address to %s for %s", $ip, $h);
+          verbose("UPDATE:","updating %s", $h);
+
+          # alexandregz: dinahosting uses subdomain into another param
+          my ($subdomain, $domain) = split(/\./, $h, 2);
+
+          # check if zone exists (Domain_Zone_UpdateTypeA) or not (Domain_Zone_AddTypeA)
+          my $command = 'Domain_Zone_UpdateTypeA';
+
+          unless($replyCheck) {
+              my $urlCheck = "https://$config{$h}{'server'}$config{$h}{'script'}";
+              $urlCheck  .= "?AUTH_USER=$config{$h}{'login'}";
+              $urlCheck  .= "&AUTH_PWD=$config{$h}{'password'}";
+              $urlCheck  .= "&domain=$domain";
+              $urlCheck  .= "&command=Domain_Zone_GetTypeA";
+              #info("urlCheck %s", $urlCheck);
+
+              $replyCheck = geturl(opt('proxy'), $urlCheck);
+              if (!defined($replyCheck) || !$replyCheck) {
+                   failed("checkzone %s: Could not connect to %s.", $h, $config{$h}{'server'});
+                   last;
+              }
+          }
+
+          $command = 'Domain_Zone_AddTypeA' unless ($replyCheck =~ /\hostname = '$subdomain'$/);
+          info("command %s - domain %s - subdomain %s - h %s", $command, $domain, $subdomain, $h);
+
+
+          # Set the URL that we're going to update
+          my $url = "https://$config{$h}{'server'}$config{$h}{'script'}";
+          $url  .= "?AUTH_USER=$config{$h}{'login'}";
+          $url  .= "&AUTH_PWD=$config{$h}{'password'}";
+          $url  .= "&hostname=$subdomain";
+          $url  .= "&domain=$domain";
+          $url  .= "&command=$command";  # new zones change to Domain_Zone_AddTypeA
+          $url  .= "&ip=";
+          $url  .= $ip            if $ip;
+          #verbose("UPDATE:","url %s", $url);
+
+          # Try to get URL
+          my $reply = geturl(opt('proxy'), $url);
+
+
+          # No response, declare as failed
+          if (!defined($reply) || !$reply) {
+               failed("updating %s: Could not connect to %s.", $h, $config{$h}{'server'});
+               last;
+          }
+          #last if !header_ok($h, $reply);
+
+          # Response found
+          if ($reply =~ /Success/i) {
+               $config{$h}{'ip'}     = $ip;
+               $config{$h}{'mtime'}  = $now;
+               $config{$h}{'status'} = 'good';
+               success("updating %s: %s: IP address set to %s", $h, 'good', $ip);
+         } else {
+               my ($statuscode, $statusmessage);
+               if($reply =~ /responseCode = (\d+)\nerrors_0_message = '(.*)'$/m) {
+                   $statuscode = $1;
+                   $statusmessage = $2;
+               }
+
+               $config{$h}{'status'} = 'failed';
+               failed("updating %s: %s - %s", $h, $statuscode, $statusmessage);
+          }
+     }
+}
+######################################################################
+## nic_dinahosting_examples
+######################################################################
+sub nic_dinahosting_examples {
+    return <<EoEXAMPLE;
+o 'dinahosting'
+
+The 'dinahosting' protocol is used by DNS services offered by www.dinahosting.com.
+
+Configuration variables applicable to the 'dinahosting' protocol are:
+  protocol=dinahosting         ##
+  server=fqdn.of.service       ## defaults to dinahosting.com
+  login=fully.qualified.host   ## login name and password registered with the service
+  password=master-password     ##
+  fully.qualified.host         ## the host registered with the service.
+
+Example ${program}.conf file entries:
+  ## single host update
+  protocol=dinahosting,                              \\
+  login=myhost.mydomain.com,                         \\
+  password=dinahosting-password                      \\
+  myhost.mydomain.com
+
+EoEXAMPLE
+}
+######################################################################
 ## nic_direktdomains_update
 ## Note: uses same features as nic_dyndns2_update, less return codes
 ######################################################################
